### ISSUE TITLE
Compile instructions (Linux)

### ISSUE THREAD

**Issue by admalledd** [kc5nra/obs-browser#1]
I would be interested in trying to help, but currently the complete lack of a clue on how to build this makes that rather hard...

[link](https://obsproject.com/forum/threads/browser-plugin.20358/#post-109057) to forum post where I note a few issues that I had while trying to build/use CEF and this plugin.

So some basic compile instructions from a system that can already build obs-studio would be nice, EG how to build CEF, the obs-browser sub-process stuff, change/set LD_LIBRARY_PATHs for out-of-band compiling/installing. Although just getting it to build _once_ would probably be enough for me to go from there and detail most of the missing information for out-of-band compiles/installs if you have issues writing instructions for those.

**Comment by kkartaltepe** [comment]
https://github.com/kkartaltepe/obs-browser Fewest changes I could make to be able to compile the plugin on Arch. Grab the newest stable CEF build from http://cefbuilds.com/ (Avoid trunk as they make breaking changes only to revert them later..). compile the libcef_wrapper inside the cef build you downloaded. `cmake ./ -DCEF_ROOT_DIR=where/cef/is` and cmake should run successfully, `make` and the plugin should be compiled in `build/`. From there it's a bit broken as https://code.google.com/p/javachromiumembedded/issues/detail?id=109 exists and still persists in chromium's trunk with the ever ironic comment.
```
   // For now, expect the data file to be alongside the executable.
    // This is sufficient while we work on unit tests, but will eventually
    // likely live in a data directory.
```

**Comment by coolacid** [comment]
Old issue is old.. but: 

```
/home/Repos/OBS/obs-browser/obs-browser/main-source.cpp: In function ‘obs_source_info create_browser_source_info()’:
/home/Repos/OBS/obs-browser/obs-browser/main-source.cpp:199:31: error: invalid conversion from ‘const char* (*)()’ to ‘const char* (*)(void*)’ [-fpermissive]
  browser_source_info.get_name = browser_source_get_name;
                               ^
CMakeFiles/obs-browser.dir/build.make:77: recipe for target 'CMakeFiles/obs-browser.dir/obs-browser/main-source.cpp.o' failed
```

**Comment by coolacid** [comment]
I see @GloriousEggroll has a bunch of work done to get this working on Linux again. Any chance we can get a PR to get to whatever point that's at?

**Comment by GloriousEggroll** [comment]
my version is not working. the best way to compile it according to the original author was to compile it with obs as a plugin, but even then it still does not work.

**Comment by stephanel** [comment]
Hello everyone,
I can configure the plugin's build by including cmake/Modules/ObsHelpers.cmake in CMakeLists.txt and then running this : 
```
cmake -DCEF_ROOT_DIR=../cef_binary_3.2704.1432.g60b3718_linux64 -DCEFWRAPPER_LIBRARY=../cef-build/libcef_dll_wrapper/libcef_dll_wrapper.a -DCMAKE_CXX_FLAGS='-std=c++0x' ../obs-browser-1.23/
```
Note the use of the `-std=c++0x` flag (without it, cef-bootstrap doesn't build). The command executes successfully, and running `make` starts the build process.

During the build, a warning is issued at 62% :
```
/home/stephane/obs-browser-1.23/obs-browser/browser-source-listener-base.cpp: In member function ‘virtual void BrowserSource::Impl::Listener::OnDraw(BrowserSurfaceHandle, int, int)’:
/home/stephane/obs-browser-1.23/obs-browser/browser-source-listener-base.cpp:42:23: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘BrowserSurfaceHandle {aka gs_texture*}’ [-Wformat=]
    "%d", surfaceHandle);
                       ^
```

And after this, the build crashes at 68% :
```
/home/stephane/obs-browser-1.23/obs-browser/browser-manager-base.cpp: In lambda function:
/home/stephane/obs-browser-1.23/obs-browser/browser-manager-base.cpp:249:50: error: cannot convert ‘char16** {aka short unsigned int**}’ to ‘wchar_t**’ for argument ‘3’ to ‘size_t os_utf8_to_wcs_ptr(const char*, size_t, wchar_t**)’
    os_utf8_to_wcs_ptr(event->text, 0, &characters);
                                                  ^
```

Since I'm not a C++ god, does anyone have an idea of what is happening ?

**Comment by kkartaltepe** [comment]
Anyone who happens upon this and is just trying to get a browser plugin for linux please see https://github.com/bazukas/obs-qtwebkit

Even if you can get this to compile there are still issues to be resolved so if your intent isnt to do some dev work you are probably in the wrong place.

**Comment by stephanel** [comment]
Maybe this issue should be closed like #10 ?

**Comment by coolacid** [comment]
I'm in favor of closing this, after a list of known issues is created for tracking items that are still outstanding.

**Comment by coolacid** [comment]
I'm currently working on this again. Currently it all builds and OBS launches however, OBS locks up while attempting to add a browser source. I've tracked this down to what appears to be CEF not either not starting a bootstrap browser, or no rendering browser - I'm not 100% sure. 

My running notes on getting this running is here: https://gist.github.com/coolacid/7f7fab7bf227620936e19985538cd1de

**Comment by coolacid** [comment]
Just to update ^ 

CefRunMessageLoop() appears to run. I put some "blog" messages sparced out in the code, and the MessageLoop appears to run, and not exit.

However,

https://github.com/kc5nra/obs-browser/blob/8da1cbcdaa50fd7dfef91f9e41fb7b2566f11389/obs-browser/browser-manager-base.cpp#L113

Never triggers, thus createdEvent signal never fires causing the hang.

**Comment by coolacid** [comment]
After spending more hours trying to figure out what's going on in Linux (This includes, getting the debug CEF builds running etc) I think I've stumbled upon a clue:

http://www.magpcss.org/ceforum/viewtopic.php?f=10&t=13043#p26128

```The largest problem is that Cocoa and X11 expect all UX-related code to execute on the main application thread. ```

If I'm reading the code correctly (and I may/maynot be cross eyed at this point) plus, not fully knowing the CEF/OBS code flow, it appears the `CefRunMessageLoop` function lives inside a separate thread? 

I'm wondering if using the `CefDoMessageLoopWork` inside a primary OBS loop would work better?

**Comment by GloriousEggroll** [comment]
hi guys, I've started to take another crack at this after a long break. I've got it compiling with OBS, and obs does load it, however when I go to add a browser source, it freeses once i hit the add button. I am working with a fork of the updated browser-source forked from this one from Osiris:
https://github.com/OsirisNL/obs-browser

my fork
https://github.com/GloriousEggroll/obs-browser

see the readme for details.
currently seems like im stuck at the same issue coolacid was having where it doesnt start/load the bootstrap browser
